crates/rpc/rpc/src/engine/jwt_layer.rs

```diff
@@ -0,0 +1,134 @@
+use http::{HeaderMap, Request, Response, StatusCode};
+
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+
+use crate::JwtSecret;
+
+/// Middleware struct that provides JWT bearer
+/// token check for incoming Http requests
+#[allow(missing_debug_implementations)]
+pub struct JwtLayer(JwtSecret);
+
+impl JwtLayer {
+    /// Creates an instance of [`JwtLayer`][crate::engine::JwtLayer].
+    pub fn new(secret: JwtSecret) -> Self {
+        Self(secret)
+    }
+}
+
+impl<S> Layer<S> for JwtLayer {
+    type Service = JwtService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        JwtService { secret: self.0.clone(), inner }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct JwtService<S> {
+    secret: JwtSecret,
+    inner: S,
+}
+
+impl<ReqBody, ResBody, S> Service<Request<ReqBody>> for JwtService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Default,
+{
+    type Response = Response<ResBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match self.get_bearer(req.headers()) {
+            Some(jwt) => match self.secret.validate(jwt) {
+                Ok(_) => ResponseFuture::future(self.inner.call(req)),
+                Err(_) => {
+                    // TODO: Log
+                    ResponseFuture::invalid_auth()
+                }
+            },
+            None => {
+                // TODO: Log
+                ResponseFuture::invalid_auth()
+            }
+        }
+    }
+}
+
+impl<S> JwtService<S> {
+    // TODO: Add full coverage
+    fn get_bearer(&self, headers: &HeaderMap) -> Option<String> {
+        headers
+            .get("authorization")
+            .map(|header| header.to_str())
+            .and_then(|result| result.ok())
+            .and_then(|auth| {
+                let prefix = "Bearer ";
+                auth.find(prefix)
+                    .map(|index| &auth[index + prefix.len()..])
+                    .map(|token| token.to_owned())
```

Comment:
this looks a bit weird, perhaps can use 
`Bearer` type https://docs.rs/tower-http/latest/src/tower_http/auth/require_authorization.rs.html#382
or mirror how this type does auth, basically tracking the expected HeaderValue

Reply:
> Bearer type https://docs.rs/tower-http/latest/src/tower_http/auth/require_authorization.rs.html#382

The code in the link makes a comparison like `"Bearer <ACTUAL>" == "Bearer <EXPECTED>"`.
We have a different requirement: extract the token value after the "Bearer" prefix to perform a deeper validation.

I simplified the code, tell me if you like it.